spec/ordinary-and-exotic-object-behaviours.html

```diff
@@ -0,0 +1,326 @@
+          1. Return *false*.
+        </emu-alg>
+      </emu-clause>
+      
+      <emu-clause id="sec-record-exotic-objects-hasproperty-p">
+        <h1>[[HasProperty]] ( _P_ )</h1>
+        <p>When the [[HasProperty]] internal method of a Record exotic object _R_ is called with property key _P_, the following steps are taken:</p>
+        <emu-alg>
+          1. Assert: IsPropertyKey(_P_) is *true*.
+          1. If Type(_P_) is a Symbol, return *false*.
+          1. Return ! RecordHasProperty(_R_, _P_).
+        </emu-alg>
+      </emu-clause>
+            
+      <emu-clause id="sec-record-exotic-objects-get-p-receiver">
+        <h1>[[Get]] ( _P_, _Receiver_ )</h1>
+        <p>When the [[Get]] internal method of a Record exotic object _R_ is called with property key _P_ and ECMAScript language value _Receiver_, the following steps are taken:</p>
+        <emu-alg>
+          1. Assert: IsPropertyKey(_P_) is *true*.
+          1. If Type(_P_) is a Symbol, return *undefined*.
+          1. Return ! RecordGet( _R_, _P_ ).
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-record-exotic-objects-set-p-receiver">
+        <h1>[[Set]] ( _P_, _Receiver_ )</h1>
+        <p>When the [[Set]] internal method of a Record exotic object _R_ is called with property key _P_ and ECMAScript language value _Receiver_, the following steps are taken:</p>
+        <emu-alg>
+          1. Return *false*.
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-record-exotic-objects-delete-p">
+        <h1>[[Delete]] ( _P_ )</h1>
+        <p>When the [[Delete]] internal method of a Record exotic object _R_ is called with property key _P_, the following steps are taken:</p>
+        <emu-alg>
+          1. Assert: IsPropertyKey(_P_) is *true*.
+          1. If Type(_P_) is a Symbol, return *true*.
+          1. Return *false*.
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-record-exotic-objects-ownpropertykeys">
+        <h1>[[OwnPropertyKeys]] ( )</h1>
+        <p>When the [[OwnPropertyKeys]] internal method of a Record exotic object _R_ is called, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _keys_ be a new empty List.
+          1. Let _rec_ be _R_.[[RecordData]].
+          1. Assert: Type(_rec_) is Record.
+          1. Let _len_ be the length of _rec_.
+          1. For each pair _p_ of _rec_.[[Fields]], do
+            1. Let _key_ be _p_.[[Key]]
+            1. Assert: Type(_key_) is String
+            1. Add _key_ to _keys_
+          1. Return _keys_
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-recordcreate" aoid="RecordCreate">
+        <h1>RecordCreate ( _value_ )</h1>
+        <p>This abstract operation RecordCreate takes a _value_ argument. It is used to specify the creation of new Record exotic objects. It performs the following steps when called:</p>
+        <emu-alg>
+            1. Assert: Type(_value_) is a Record.
```

Comment:
why would RecordCreate take a Record? This seems like more the domain of https://tc39.es/ecma262/#sec-toobject than of its own abstract op

Reply:
I was a little mystified by the naming convention at first, but see analogous operations like https://tc39.es/ecma262/#sec-stringcreate and https://tc39.es/ecma262/#sec-modulenamespacecreate

Reply:
hmm, strange that String has it but Number, Boolean, etc don't (objects having a FooCreate makes sense to me; but primitives don't). It feels like the precedent is more for primitives not to have one.

Reply:
I guess what distinguishes String is that its wrapper is exotic, like Record. Maybe we should continue this discussion in an editorial issue in ecma262? I want to make sure we have clear text somewhere or other to explain how the wrapper is set up, and I think it should match String, but I am fine with moving to some other way of organizing the text.

Reply:
I put those functions here as a temporary location, they are not meant to stay there

Reply:
Please open an issue if you think this should still be changed @ljharb , we can discuss the choices in more detail.